cloud/scope/azureippool.go

```diff
@@ -0,0 +1,348 @@
+		AzureIPPool: params.AzureIPPool,
+		patchHelper:  helper,
+		//Lock: params.Lock,
+	}, nil
+}
+
+// ClusterScope defines the basic context for an actuator to operate upon.
+type AzureIPPoolScope struct {
+	logr.Logger
+	Client      client.Client
+	patchHelper *patch.Helper
+	//Lock sync.Mutex
+	AzureIPPool *infrav1.AzureIPPool
+}
+
+// PatchObject persists the cluster configuration and status.
+func (s *AzureIPPoolScope) PatchObject(ctx context.Context) error {
+	return s.patchHelper.Patch(
+		ctx,
+		s.AzureIPPool,
+		patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+			clusterv1.ReadyCondition,
+			infrav1.NetworkInfrastructureReadyCondition,
+		}})
+}
+
+// Close closes the current scope persisting the cluster configuration and status.
+func (s *AzureIPPoolScope) Close(ctx context.Context) error {
+	return s.patchHelper.Patch(ctx, s.AzureIPPool)
+}
+
+func (s *AzureIPPoolScope) GetIPPoolObj(ctx context.Context, namespace string, ipPoolName string) (*infrav1.AzureIPPool, error) {
+	azureIPPoolName := client.ObjectKey{
+		Namespace: namespace,
+		Name:      ipPoolName,
+	}
+	azureippool := &infrav1.AzureIPPool{}
+	if err := s.Client.Get(ctx, azureIPPoolName, azureippool); err != nil {
+	return azureippool, errors.Wrapf(err, "Failed to retrieve Azure IP pool")
+	}
+	return azureippool, nil
+}
+
+/*Reconcile IP pool gets the IP from IP pool, adds it to allocated IP pool and updates the IP pool. Machine IP is "" for worker VMs and
+equal to APIServerIP for control plane VM*/
+func (s *AzureIPPoolScope) ReconcileIPs(ctx context.Context, machineScope *MachineScope, ippoolname string, machineIP string) error {
+	lock.Lock()
+	machineScope.Info(fmt.Sprintf("Locking reconcile IP for machine %s", machineScope.Name()))
+	defer func() {
+		machineScope.Info(fmt.Sprintf("Unlocking reconcile IP for machine %s", machineScope.Name()))
+     lock.Unlock()   
+    }()
+	
+	azureIPPool, err := s.GetIPPoolObj(ctx, "default", ippoolname)
+    if err!=nil {
+        return errors.Wrapf(err, "Failed to retrieve Azure IP pool")
+    }
+	
+	if azureIPPool.Spec.IPPools == nil {
+		return errors.New("IP Pool cannot be nil")
+	}
+
+	for index, _ := range(azureIPPool.Spec.IPPools) {
+		ippool := azureIPPool.Spec.IPPools[index]        
+		networkInterface := BuildNetworkInterfaceSpec(machineScope, ippool, ippool.Name)
+		if len(ippool.IPs) == 0 && ippool.Name == "mgmt-nic-ip-pool" {
+			return errors.New(fmt.Sprintf("IPs cannot be nil for mgmt-nic-ip-pool"))
+		}
+        if len(ippool.IPs) > 0 {
+			allocatedIPsMap, err := BuildAllocatedIPsMap(azureIPPool, ippool)
+			if err != nil {
+				return errors.Wrapf(err, fmt.Sprintf("Failed to reconcile IP for AzureMachine %s", machineScope.AzureMachine.Name))
+			}
+			machineScope.Info(fmt.Sprintf("allocatedIPsMap %v", allocatedIPsMap))
+			allocatedIP := ""
+			// machine IP is "" for worker VM and equal to APIServerIP for control plane VM
+			if ippool.Name != "mgmt-nic-ip-pool" {
+				machineIP = ""
```

Comment:
Shouldn't the caller already set this to "" if it is a worker node.

Reply:
Reply at https://github.com/niachary/cluster-api-provider-azure/pull/5/files#r684344631